File: sdk-android-sample/src/main/java/com/xyfindables/sdk/sample/XYDeviceAdapter.java
package com.xyfindables.sdk.sample;

import android.app.Activity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import com.xyfindables.sdk.XYDevice;
import com.xyfindables.sdk.XYSmartScan;

import java.util.ArrayList;

public class XYDeviceAdapter extends BaseAdapter {

    private static String TAG = XYDeviceAdapter.class.getSimpleName();

    private Activity _activity;
    private ArrayList<XYDevice> _devices;

    public XYDeviceAdapter(Activity activity) {
        _activity = activity;
        _devices = new ArrayList<XYDevice>();
        XYSmartScan.instance.addListener(TAG, new XYDevice.Listener() {
            @Override
            public void entered(final XYDevice device) {
                _activity.runOnUiThread (new Thread(new Runnable() {
                    public void run() {
                        _devices.add(device);
                        notifyDataSetChanged();
                    }
                }));
            }

            @Override
            public void exited(final XYDevice device) {
                _activity.runOnUiThread (new Thread(new Runnable() {
                    public void run() {
                        _devices.remove(device);
                        notifyDataSetChanged();
                    }
                }));
            }

            @Override
            public void detected(XYDevice device) {
            }

            @Override
            public void buttonPressed(XYDevice device, XYDevice.ButtonType buttonType) {
            }

            @Override
            public void buttonRecentlyPressed(XYDevice device, XYDevice.ButtonType buttonType) {
            }

            @Override
            public void statusChanged(XYSmartScan.Status status) {
            }

            @Override
            public void updated(XYDevice device) {

            }

            @Override
            public void connectionStateChanged(XYDevice device, int newState) {

            }

            @Override
            public void readRemoteRssi(XYDevice device, int rssi) {

            }

        });
    }

    @Override
    public int getCount() {
        return _devices.size();
    }

    @Override
    public Object getItem(int position) {
        return _devices.get(position);
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).hashCode();
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if (convertView == null) {
            convertView = _activity.getLayoutInflater().inflate(R.layout.device_item, parent, false);
        }
        ((XYDeviceItemView)convertView).setDevice((XYDevice)getItem(position));

        return convertView;
    }
}

File: sdk-android-sample/src/main/java/com/xyfindables/sdk/sample/XYDeviceAdapter.kt
package com.xyfindables.sdk.sample

import android.app.Activity
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.xyfindables.sdk.devices.XYBluetoothDevice

import com.xyfindables.sdk.scanner.XYFilteredSmartScan
import com.xyfindables.ui.ui

import java.util.ArrayList

class XYDeviceAdapter(private val activity: Activity) : BaseAdapter() {
    private var devices: List<Pair<Int, XYBluetoothDevice>>

    val scanner : XYFilteredSmartScan
        get() {
            return (activity.applicationContext as XYApplication).scanner
        }

    val smartScannerListener = object : XYFilteredSmartScan.Listener() {
        override fun entered(device: XYBluetoothDevice) {
            ui {
                devices = scanner.devices.toList()
                notifyDataSetChanged()
            }
        }

        override fun exited(device: XYBluetoothDevice) {
            ui {
                devices = scanner.devices.toList()
                notifyDataSetChanged()
            }
        }
    }

    init {
        devices = ArrayList()
        scanner.addListener(TAG, smartScannerListener)
    }

    override fun getCount(): Int {
        return devices.size
    }

    override fun getItem(position: Int): Any {
        return devices[position].second
    }

    override fun getItemId(position: Int): Long {
        return devices[position].second.address.hashCode().toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        if (view == null) {
            view = activity.layoutInflater.inflate(R.layout.device_item, parent, false)
        }
        (view as XYDeviceItemView).setDevice(getItem(position) as XYBluetoothDevice)

        return view
    }

    companion object {

        private val TAG = XYDeviceAdapter::class.java.simpleName
    }
}
